lib/CodeGen/CGOpenCLRuntime.cpp

```diff
@@ -77,9 +77,9 @@ llvm::Type *CGOpenCLRuntime::convertOpenCLSpecificType(const Type *T) {
     return llvm::PointerType::get(llvm::StructType::create(
                            Ctx, "opencl.image3d_t"), ImgAddrSpc);
   case BuiltinType::OCLSampler:
-    if (CGM.getLangOpts().CLKeepSamplerType)
-      return llvm::StructType::create(Ctx, llvm::IntegerType::get(Ctx, 32),
-                                      "opencl.sampler_t");
+    if (CGM.getLangOpts().CLSamplerOpaque)
+      return llvm::PointerType::get(llvm::StructType::create(
+                           Ctx, "spirv.Sampler"), 0);
```

Comment:
This helps backends for optimization since they know the samplers do not change.